physics/dynamic_frame_body.hpp

```diff
@@ -38,5 +40,40 @@ DynamicFrame<InertialFrame, ThisFrame>::FrenetFrame(
                  binormal.coordinates()).Transpose());
 }
 
+template<typename InertialFrame, typename ThisFrame>
+std::unique_ptr<DynamicFrame<InertialFrame, ThisFrame>>
+DynamicFrame<InertialFrame, ThisFrame>::ReadFromMessage(
+    not_null<Ephemeris<InertialFrame> const*> const ephemeris,
+    serialization::DynamicFrame const& message) {
+  std::unique_ptr<DynamicFrame> result;
+  int extensions_found = 0;
+  // TODO(egg): See if the reset/release pairs could be avoided by smart
+  // conversions of |not_null|.
+  if (message.HasExtension(serialization::BarycentricRotatingDynamicFrame::
+                               barycentric_rotating_dynamic_frame)) {
+    ++extensions_found;
+    result.reset(BarycentricRotatingDynamicFrame<InertialFrame, ThisFrame>::
+        ReadFromMessage(
+            ephemeris,
+            message.GetExtension(
+              serialization::BarycentricRotatingDynamicFrame::
+                  barycentric_rotating_dynamic_frame)).release());
+  }
+  if (message.HasExtension(serialization::BodyCentredNonRotatingDynamicFrame::
+                               body_centred_non_rotating_dynamic_frame)) {
+    ++extensions_found;
+    result.reset(BodyCentredNonRotatingDynamicFrame<InertialFrame, ThisFrame>::
+        ReadFromMessage(
+            ephemeris,
+            message.GetExtension(
+                serialization::BodyCentredNonRotatingDynamicFrame::
+                    body_centred_non_rotating_dynamic_frame)).release());
+  }
+  CHECK_GE(1, extensions_found) << message.DebugString();
```

Comment:
Could we write this `CHECK_LE(extensions_found, 1)` instead?
This sounds like a lower bound on `1` rather than an upper bound on `extensions_found`.